include/zephyr/device.h

```diff
@@ -923,7 +926,634 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 	Z_INIT_ENTRY_DEFINE(DEVICE_NAME_GET(dev_name), init_fn,		\
 		(&DEVICE_NAME_GET(dev_name)), level, prio)
 
-#if CONFIG_HAS_DTS
+/**
+ * @brief Multi API Device Model APIs
+ * @defgroup device_model_multi_api Multi API Device Model APIs
+ *
+ * @note The multi API devices do not support device handles, and can not
+ * be looked up using device_get_binding().
+ *
+ * @{
+ */
+
+/**
+ * @def DEVICE_API_SPEC
+ *
+ * @brief Named tuple containing device API parameters used during device
+ * definition.
+ *
+ * @param api_ptr Pointer to the device's API structure.
+ *
+ * @param api_type Lowercase shorthand type of the device's API structure
+ * (uart, spi, i2c, ...).
+ */
+#define DEVICE_API_SPEC(api_ptr, api_type) \
+	(api_ptr, api_type)
+
+/**
+ * @def DEVICE_API_DEFINE
+ *
+ * @brief Define required <tt>struct device</tt> objects for device and
+ * set it up for boot time initialization.
+ *
+ * @details A <tt>struct device</tt> is created for each specified API.
+ *
+ * The following example shows how to define three <tt>struct device</tt>
+ * objects for a device which implements the uart, i2c and spi APIs.
+ *
+ * DEVICE_API_DEFINE(dev_name, drv_name, ...,
+ *     DEVICE_API_SPEC(&uart_ptr, uart),
+ *     DEVICE_API_SPEC(&i2c_ptr, i2c),
+ *     DEVICE_API_SPEC(&spi_ptr, spi));
+ *
+ * See DEVICE_API_GET() to get a pointer to a <tt>struct device</tt>
+ * defined using this macro.
+ *
+ * @param dev_name A unique token which is used as the base name for
+ * the <tt>struct device</tt> objects C identifiers. The API type of
+ * the <tt>struct device</tt> object is appended to the C identifier.
+ *
+ * @param drv_name A string base name for the <tt>struct device</tt>
+ * objects, which will be stored in the <tt>struct device</tt>
+ * object's @p name field. The API type of the <tt>struct device</tt>
+ * is appended to the name.
+ *
+ * @param init_fn Pointer to the device's initialization function,
+ * which will be run by the kernel during system initialization.
+ *
+ * @param pm_dev_ptr Pointer to the device's power management
+ * resources, a <tt>struct pm_device</tt>, which will be stored in all
+ * <tt>struct device</tt> objects @p pm field. Set to NULL if the
+ * device does not use PM.
+ *
+ * @param data_ptr Pointer to the device's private mutable data, which
+ * is stored in the <tt>struct device</tt> objects @p data field.
+ *
+ * @param cfg_ptr Pointer to the device's private constant data, which
+ * is stored in the <tt>struct device</tt> objects @p config field.
+ *
+ * @param init_level The device's initialization level. See SYS_INIT() for
+ * details.
+ *
+ * @param init_prio The device's priority within its initialization level.
+ * See SYS_INIT() for details.
+ *
+ * @param api_spec The device's primary API specification, defined using
+ * DEVICE_API_SPEC().
+ *
+ * @param ... List of further unique API specifications.
+ */
+#define DEVICE_API_DEFINE(dev_name, drv_name, init_fn, pm_dev_ptr, data_ptr,        \
```

Comment:
No, not a great name. 

Actually all the macro created here with "DEVICE_API" should have their name revised to something more relevant to what is really does.

Current proposed name can be understood in different ways:
- Is this defining a device API (and thus **not** a device per se)?
- Is this defining a device **with** an API? But then, that can be already done with DEVICE_DEFINE()

Let's take for instance: DEVICE_WITH_MULTIPLE_API_DEFINE(), such name offers no ambiguity. Though perhaps it can be shortened: WITH_MULTIPLE_API would become WMAPI?

Reply:
It is a replacement for DEVICE_DEFINE and DEVICE_DT_DEFINE, the _API_ is simply to distinguish them in the transition period. DEVICE_DT_MULTI_API_DEFINE would be fine for now.

Reply:
An important note here, the new model supports defining more than one API, but it is not limited to multi API devices. You have to define at least one API, but single API devices are also defined using DEVICE_API_DEFINE()

I will rename the functions to _M_API_, MAPI is shorthand for "Messaging Application Programming Interface"...

Reply:
Nevermind, i am changing it to NEW instead of API, that is very descriptive :)